contracts/ChainlinkPriceFeedV3.sol

```diff
@@ -0,0 +1,213 @@
+import { BlockContext } from "./base/BlockContext.sol";
+import { CachedTwap } from "./twap/CachedTwap.sol";
+
+contract ChainlinkPriceFeedV3 is IChainlinkPriceFeedV3, BlockContext, CachedTwap {
+    using SafeMath for uint256;
+    using Address for address;
+
+    //
+    // STATE
+    //
+
+    uint24 private constant _ONE_HUNDRED_PERCENT_RATIO = 1e6;
+    uint8 internal immutable _decimals;
+    uint24 internal immutable _maxOutlierDeviationRatio;
+    uint256 internal immutable _outlierCoolDownPeriod;
+    uint256 internal immutable _timeout;
+    uint256 internal _lastValidPrice;
+    uint256 internal _lastValidTimestamp;
+    AggregatorV3Interface internal immutable _aggregator;
+
+    //
+    // EXTERNAL NON-VIEW
+    //
+
+    constructor(
+        AggregatorV3Interface aggregator,
+        uint256 timeout,
+        uint24 maxOutlierDeviationRatio,
+        uint256 outlierCoolDownPeriod,
+        uint80 twapInterval
+    ) CachedTwap(twapInterval) {
+        // CPF_ANC: Aggregator is not contract
+        require(address(aggregator).isContract(), "CPF_ANC");
+        _aggregator = aggregator;
+
+        // CPF_IMODR: Invalid maxOutlierDeviationRatio
+        require(maxOutlierDeviationRatio < _ONE_HUNDRED_PERCENT_RATIO, "CPF_IMODR");
+        _maxOutlierDeviationRatio = maxOutlierDeviationRatio;
+
+        _outlierCoolDownPeriod = outlierCoolDownPeriod;
+        _timeout = timeout;
+        _decimals = aggregator.decimals();
+    }
+
+    /// @inheritdoc IChainlinkPriceFeedV3
+    function cacheTwap(uint256 interval) external override {
+        _cachePrice();
+
+        if (interval != 0) {
+            _cacheTwap(interval, _lastValidPrice, _lastValidTimestamp);
+        }
+    }
+
+    //
+    // EXTERNAL VIEW
+    //
+
+    function getLastValidPrice() external view override returns (uint256) {
+        return _lastValidPrice;
+    }
+
+    function getLastValidTimestamp() external view override returns (uint256) {
+        return _lastValidTimestamp;
+    }
+
+    /// @inheritdoc IChainlinkPriceFeedV3
+    function getCachedTwap(uint256 interval) external view override returns (uint256) {
+        (uint256 latestValidPrice, uint256 latestValidTime) = _getCachePrice();
+
+        if (interval == 0) {
+            return latestValidPrice;
+        }
+
+        return _getCachedTwap(interval, latestValidPrice, latestValidTime);
+    }
+
+    function getAggregator() external view override returns (address) {
+        return address(_aggregator);
+    }
+
+    function decimals() external view override returns (uint8) {
+        return _decimals;
+    }
+
+    function isTimedOut() external view override returns (bool) {
+        return _lastValidTimestamp.add(_timeout) > _blockTimestamp();
+    }
+
+    //
+    // INTERNAL
+    //
+
+    function _cachePrice() internal {
+        ChainlinkResponse memory response = _getChainlinkResponse();
+        if (_lastValidTimestamp != 0 && _lastValidTimestamp == response.updatedAt) {
```

Comment:
Is it possible that multiple price data have the same `updatedAt` but different `answer`? So we will only cache the first `answer`. I guess that's fine.

Reply:
yep!